src/tpm2tss_engine.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2020 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <openssl/ssl.h>
+#include <openssl/ui.h>
+#include <openssl/engine.h>
+#include <stdio.h>
+#include <string.h>
+#include "log.h"
+#include "tpm2tss_engine.h"
+
+static ENGINE *
+enftun_tpm2tss_engine_init()
+{
+    ENGINE *eng = NULL;
+    const char *eng_name = "tpm2tss";
+
+    /* Init the engine */
+    ENGINE_load_builtin_engines();
+
+    eng = ENGINE_by_id(eng_name);
+    if (!eng) {
+        fprintf(stderr, "%s ERR: Could not find engine\n", __func__);
+        goto out;
+    }
+
+    if (!ENGINE_init(eng)) {
+        fprintf(stderr, "%s ERR: Could not tpm2tss engine\n", __func__);
+        ENGINE_free(eng);
+        goto out;
+    }
+
+    if (!ENGINE_set_default(eng, ENGINE_METHOD_ALL)) {
```

Comment:
This gets into a version issue. `OpenSSL 1.1.1` does not have `ENGINE_METHOD_ECDSA` (it only has DSA, which prior versions also have). It doesn't look like we need to set any defaults here since all the engine calls explicitly name which one to use.

Reply:
We don't have to call this at all?  I guess because the private key is linked to the engine, any calls to sign using that key will also use the engine.  Great.

Reply:
We might to still call `ENGINE_set_default_RAND(eng)`, to be sure that the TPM is used for randomness, rather than whatever happens to be configured on that machine.  

For the router card, it doesn't matter. We've configured Linux to use the TPM to seed that random pool.  But other people using `enftun` directly might not have done that.

Reply:
I could see that happening with something like RAND where more than one method will now be enabled. Added.